language/ocr_nl/main.py

```diff
@@ -0,0 +1,363 @@
+#!/usr/bin/env python
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This example uses the Google Cloud Vision API to detect text in images, then
+analyzes that text using the Google Cloud Natural Language API to detect
+entities in the text. It stores the detected entity information in an sqlite3
+database, which may then be queried.
+
+After this script has analyzed a directory of images, it outputs some
+information on the images' entities to STDOUT. You can also further query
+the generated sqlite3 database; see the README for more information.
+
+Run the script on a directory of images to do the analysis, E.g.:
+    $ python main.py --input_directory=<path-to-image-directory>
+
+You can try this on a sample directory of images:
+    $ curl -O http://storage.googleapis.com/python-docs-samples-tests/language/ocr_nl-images.zip
+    $ unzip ocr_nl-images.zip
+    $ python main.py --input_directory=images/
+
+"""  # noqa
+
+import argparse
+import base64
+import contextlib
+import logging
+import os
+import re
+import sqlite3
+import sys
+import time
+
+from googleapiclient import discovery
+from googleapiclient import errors
+import httplib2
+from oauth2client.client import GoogleCredentials
+
+BATCH_SIZE = 10
+
+
+class VisionApi(object):
+    """Construct and use the Cloud Vision API service."""
+
+    def __init__(self):
+        credentials = GoogleCredentials.get_application_default()
+        self.service = discovery.build('vision', 'v1', credentials=credentials)
+
+    def detect_text(self, input_filenames, num_retries=3, max_results=6):
+        """Uses the Vision API to detect text in the given file."""
+        batch_request = []
+        for filename in input_filenames:
+            request = {
+                'image': {},
+                'features': [{
+                    'type': 'TEXT_DETECTION',
+                    'maxResults': max_results,
+                }]
+            }
+
+            # Accept both files in cloud storage, as well as local files.
+            if re.match(r'gs://', filename):
```

Comment:
Good point.
I think the `re` is vestigial for when I was using `https?://` as the protocol to check for. But then I realized I could just pass `gs://` uris directly..